src/Microsoft.ML.Data/Evaluators/RankingEvaluator.cs

```diff
@@ -948,41 +953,41 @@ private protected override IEnumerable<string> GetPerInstanceColumnsToSave(RoleM
 
     internal static class RankingUtils
     {
-        private static readonly object _lock = new object();
-        private static volatile Double[] _discountMap;
-        private static volatile int _maxTruncationLevel = 0;
+        // Truncation levels are typically less than 100. So we maintain a fixed discount map of size 100
+        // If truncation level greater than 100 is required, we build a new one and return that.
+        private const int FixedDiscountMapSize = 100;
+        private static Double[] _discountMapFixed;
 
-        public static Double[] DiscountMap
+        private static Double[] GetDiscountMapCore(int truncationLevel)
         {
-            get
-            {
-                return _discountMap;
-            }
+            var discountMap = new Double[FixedDiscountMapSize];
+
+            for (int i = 0; i < discountMap.Length; i++)
+                discountMap[i] = 1 / Math.Log(2 + i);
+
+            return discountMap;
         }
-        /// <summary>
-        /// Reallocates discountMap for the largest truncationLevel seen so far
-        /// </summary>
-        public static void SetTruncationLevel(int truncationLevel)
-        {
-            lock (_lock) {
-                if (truncationLevel > _maxTruncationLevel)
-                {
-                    _maxTruncationLevel = truncationLevel;
 
-                    var discountMap = new Double[_maxTruncationLevel];
-                    for (int i = 0; i < discountMap.Length; i++)
-                    {
-                        discountMap[i] = 1 / Math.Log(2 + i);
-                    }
-                    _discountMap = discountMap;
-                }
+        public static Double[] GetDiscountMap(int truncationLevel)
+        {
+            var discountMap = _discountMapFixed;
+            if (discountMap == null)
+            {
+                discountMap = GetDiscountMapCore(truncationLevel);
```

Comment:
First check if the requested level is small enough for the “fixed” map, if it is small enough, use the cached one. If the cached one hasn’t been created yet, then create it using ‘FixedDiscountMapSize’.